ocsp/ocsp_nonblock/ocsp_nonblock.c

```diff
@@ -0,0 +1,301 @@
+"30pP0so32wVtqAKZOdtgYyQs1WXEgOVouGkecgdKv2pMyWa6TVjMNnMxCwqq4MRG\n"
+"R5thr5l5tg20zvpGM7bE/VuYegTSqQyaF6arUpjpOX7xclfERZ1RUOh1EHHnH4gf\n"
+"l7eOUXh950nbb3bjp2bUF1CjsnveJI1UfqcUrp3Tuoh7ScT1gEiJ82qGsVtyq3AU\n"
+"FvKz0TJH0ipymA==\n"
+"-----END CERTIFICATE-----";
+
+
+#ifndef HTTP_SCRATCH_BUFFER_SIZE
+    #define HTTP_SCRATCH_BUFFER_SIZE 512
+#endif
+#ifndef MAX_URL_ITEM_SIZE
+    #define MAX_URL_ITEM_SIZE   80
+#endif
+#ifndef DEFAULT_TIMEOUT_SEC
+    #define DEFAULT_TIMEOUT_SEC 0 /* no timeout */
+#endif
+
+static int io_timeout_sec = DEFAULT_TIMEOUT_SEC;
+
+/* Return size of the OCSP response or negative for error */
+static int OcspLookupNonBlockCb(void* ctx, const char* url, int urlSz,
+                        byte* ocspReqBuf, int ocspReqSz, byte** ocspRespBuf)
+{
+    SOCKET_T sfd = SOCKET_INVALID;
+    word16   port;
+    int      ret = -1;
+    char     path[MAX_URL_ITEM_SIZE];
+    char     domainName[MAX_URL_ITEM_SIZE];
+    int      nonBlockCnt = 0;
+
+    if (ocspReqBuf == NULL || ocspReqSz == 0) {
+        printf("OCSP request is required for lookup\n");
+    }
+    else if (ocspRespBuf == NULL) {
+        printf("Cannot save OCSP response\n");
+    }
+    else if (wolfIO_DecodeUrl(url, urlSz, domainName, path, &port) < 0) {
+        printf("Unable to decode OCSP URL\n");
+    }
+    else {
+        /* Note: This is free'd in OcspRespFreeCb callback */
+        int   httpBufSz = HTTP_SCRATCH_BUFFER_SIZE;
+        byte* httpBuf   = (byte*)XMALLOC(httpBufSz, NULL, DYNAMIC_TYPE_OCSP);
+
+        printf("OCSP Lookup:\n");
+        printf("\tURL: %s\n", url);
+        printf("\tDomain: %s\n", domainName);
+        printf("\tPath: %s\n", path);
+        printf("\tPort: %d\n", port);
+
+        if (httpBuf == NULL) {
+            printf("Unable to create OCSP response buffer\n");
+        }
+        else {
+            httpBufSz = wolfIO_HttpBuildRequestOcsp(domainName, path, ocspReqSz,
+                                                            httpBuf, httpBufSz);
+
+            ret = wolfIO_TcpConnect(&sfd, domainName, port, io_timeout_sec);
+            if (ret == 0) {
+            #if defined(WOLFSSL_NONBLOCK_OCSP) && defined(HAVE_IO_TIMEOUT)
+                wolfIO_SetBlockingMode(sfd, 1); /* non-blocking */
+            #endif
+
+                if (wolfIO_Send(sfd, (char*)httpBuf, httpBufSz, 0) !=
+                                                                    httpBufSz) {
+                    printf("OCSP http request failed\n");
+                }
+                else if (wolfIO_Send(sfd, (char*)ocspReqBuf, ocspReqSz, 0) !=
+                                                                    ocspReqSz) {
+                    printf("OCSP ocsp request failed\n");
+                }
+                else {
+                    do {
+                        ret = wolfIO_HttpProcessResponseOcsp(sfd, ocspRespBuf, 
+                                        httpBuf, HTTP_SCRATCH_BUFFER_SIZE, ctx);
```

Comment:
The `ctx` is not the heap hint. It is in `EmbedOcspLookup` but not for user callbacks.